core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,136 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Queues;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Types.StructType;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.PartitionMap;
+import org.apache.iceberg.util.PartitionUtil;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+
+public class PartitionStatsUtil {
+
+  private PartitionStatsUtil() {}
+
+  /**
+   * Computes the partition stats for the given snapshot of the table.
+   *
+   * @param table the table for which partition stats to be computed.
+   * @param snapshot the snapshot for which partition stats is computed.
+   * @return the collection of {@link PartitionStats}
+   */
+  public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot) {
+    Preconditions.checkArgument(table != null, "table cannot be null");
+    Preconditions.checkArgument(snapshot != null, "snapshot cannot be null");
+
+    StructType partitionType = Partitioning.partitionType(table);
+    if (partitionType.fields().isEmpty()) {
+      throw new UnsupportedOperationException(
+          "Computing partition stats for an unpartitioned table");
+    }
+
+    List<ManifestFile> manifests = snapshot.allManifests(table.io());
+    Queue<PartitionMap<PartitionStats>> statsByManifest = Queues.newConcurrentLinkedQueue();
+    Tasks.foreach(manifests)
+        .stopOnFailure()
+        .throwFailureWhenFinished()
+        .executeWith(ThreadPools.getWorkerPool())
+        .run(manifest -> statsByManifest.add(collectStats(table, manifest, partitionType)));
+
+    return mergeStats(statsByManifest, table.specs());
+  }
+
+  /**
+   * Sorts the {@link PartitionStats} based on the partition data.
+   *
+   * @param stats collection of {@link PartitionStats} which needs to be sorted.
+   * @param partitionType unified partition schema.
+   * @return the list of {@link PartitionStats}
+   */
+  public static List<PartitionStats> sortStats(
+      Collection<PartitionStats> stats, StructType partitionType) {
+    List<PartitionStats> entries = Lists.newArrayList(stats.iterator());
+    entries.sort(
+        Comparator.comparing(PartitionStats::partition, Comparators.forType(partitionType)));
+    return entries;
+  }
+
+  private static PartitionMap<PartitionStats> collectStats(
+      Table table, ManifestFile manifest, StructType partitionType) {
+    try (ManifestReader<?> reader = openManifest(table, manifest)) {
+      PartitionMap<PartitionStats> statsMap = PartitionMap.create(table.specs());
+
+      for (ManifestEntry<?> entry : reader.entries()) {
+        ContentFile<?> file = entry.file();
+        PartitionSpec spec = table.specs().get(file.specId());
```

Comment:
All entries for a manifest belong to the same partition spec. Technically, we don't need to call this per each entry and can move it outside of the loop and put `spec` and `specId` next to the stats map.

We can even call `reader.spec()` directly to get the spec.